client/writer.go

```diff
@@ -0,0 +1,376 @@
+				}}
+			encodeErr = encoder.EncodeMessage(msg)
+			if encodeErr != nil {
+				break
+			}
+
+			// If the buffer isn't big enough continue to the next iteration.
+			if sizeAfter := encoder.Len(); sizeAfter < w.flushSize {
+				continue
+			}
+
+			// Otherwise we enqueue the current buffer.
+			buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
+
+			// Unlock the encoder before we enqueue the old buffer to ensure other
+			// goroutines have an oppurtunity to encode metrics while larger timer
+			// batches are being encoded.
+			encoder.Unlock()
+
+			enqueueErr = w.enqueueBuffer(buffer)
+
+			// Re-lock the encoder and update variables since the encoder's buffer
+			// may have been updated.
+			encoder.Lock()
+			sizeBefore = encoder.Len()
+
+			if enqueueErr != nil {
+				break
+			}
+		}
+	case unaggregated.GaugeType:
+		msg := encoding.UnaggregatedMessageUnion{
+			Type: encoding.GaugeWithMetadatasType,
+			GaugeWithMetadatas: unaggregated.GaugeWithMetadatas{
+				Gauge:           metric.Gauge(),
+				StagedMetadatas: metadatas,
+			}}
+		encodeErr = encoder.EncodeMessage(msg)
+	default:
+		encodeErr = errUnrecognizedMetricType
+	}
+
+	if encodeErr != nil {
+		w.log.WithFields(
+			log.NewField("metric", metric),
+			log.NewField("metadatas", metadatas),
+			log.NewErrField(encodeErr),
+		).Error("encode untimed metric error")
+		// Rewind buffer and clear out the encoder error.
+		encoder.Truncate(sizeBefore)
+		encoder.Unlock()
+		return encodeErr
+	}
+
+	if enqueueErr != nil {
+		encoder.Unlock()
+		return enqueueErr
+	}
+
+	// If the buffer size is not big enough, do nothing.
+	if sizeAfter := encoder.Len(); sizeAfter < w.flushSize {
+		encoder.Unlock()
+		return nil
+	}
+
+	// Otherwise we enqueue the current buffer.
+	buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
+	encoder.Unlock()
+	return w.enqueueBuffer(buffer)
+}
+
+// prepareEnqueueBufferWithLock prepares the writer to enqueue a
+// buffer onto its instance queue. It gets a new buffer from pool,
+// copies the bytes exceeding sizeBefore to it, resets the encoder
+// with the new buffer, and returns the old buffer.
+func (w *writer) prepareEnqueueBufferWithLock(
+	encoder *lockedEncoder,
+	sizeBefore int,
+) protobuf.Buffer {
+	buf := encoder.Relinquish()
+	if sizeBefore == 0 {
+		// If a single write causes the buffer to exceed the flush size,
+		// reset and send the buffer as is.
+		return buf
+	}
+	// Otherwise we reset the buffer and copy the bytes exceeding sizeBefore,
+	// and return the old buffer.
+	encoder.Reset(buf.Bytes()[sizeBefore:])
+	buf.Truncate(sizeBefore)
+	return buf
+}
+
+func (w *writer) flushWithLock() error {
+	multiErr := xerrors.NewMultiError()
+	for _, encoder := range w.encodersByShard {
+		encoder.Lock()
+		if encoder.Len() == 0 {
+			encoder.Unlock()
+			continue
+		}
+		buffer := encoder.Relinquish()
+		encoder.Unlock()
+		if err := w.enqueueBuffer(buffer); err != nil {
+			multiErr = multiErr.Add(err)
```

Comment:
does user handle the err from just one encoder?

Reply:
right, but the buffer that failed to be enqueued here will just be discarded I guess?

Reply:
Correct, if the queue is full, the buffer will be discarded during the enqueue operation.